Spinctrl fixes

thesiv opened the pull request.

thesiv committed:
Improve the best size determination for wxSpinCtrl

Use the range and the base to determine the widest value string and use
it to calculate the best size.

thesiv committed:
Prettify wxGridCellNumberEditor placement in the grid

Use the best height and don't let the editor be smaller then min size.
Also align center vertically.

thesiv committed:
Allow process Enter and Tab in wxGridCellNumberEditor

Add style flags to allow Tab and Enter works properly for the grid number
efitor with ranges (which uses wxSpinCtrl).

vadz commented on src/generic/grideditors.cpp:
```diff
@@ -703,6 +707,40 @@ void wxGridCellNumberEditor::Create(wxWindow* parent,
+        const int width = wxMax(r.GetWidth(), bestSize.GetWidth());
+        const int height = bestSize.GetHeight() > 0 ?
+                           bestSize.GetHeight() :
+                           r.GetHeight();
```
This seems to be unused. It was probably meant to be used in the line just below?

vadz commented on src/generic/grideditors.cpp:
```diff
@@ -703,6 +707,40 @@ void wxGridCellNumberEditor::Create(wxWindow* parent,
+                           bestSize.GetHeight() :
+                           r.GetHeight();
+
+        wxRect rect(r.GetLeft(), r.GetTop(), width, bestSize.GetHeight());
```
```suggestion
        wxRect rect(r.GetPosition(), wxSize(width, height));
```

Actually I'd also use a `wxSize` instead of separate `width` and `height`, e.g.

```cpp
wxSize size = Spin()->GetBestSize();

// Extend the control to fill the entire cell horizontally.
if ( size.x < r.GetWidth() )
    size.x = r.GetWidth();

// Ensure it uses a reasonable height even if wxSpinCtrl::GetBestSize()
// didn't return anything useful.
if ( size.y <= 0 )
    size.y = r.GetHeight();
```

vadz commented on src/generic/grideditors.cpp:
```diff
@@ -703,6 +707,40 @@ void wxGridCellNumberEditor::Create(wxWindow* parent,
+
+        wxRect rect(r.GetLeft(), r.GetTop(), width, bestSize.GetHeight());
+
+        // If edit not the top most cell then align center.
```
```suggestion
        // If possible, i.e. if we're not editing the topmost or leftmost cell, center the control rectangle in the cell.
```

BTW, it would be great to rename `r` and `rect` to `rectCell` and `rectSpin` or something like this to make it immediately clear which one is which -- which is not at all the case now.

vadz commented on src/msw/spinctrl.cpp:
```diff
@@ -352,33 +371,32 @@ bool wxSpinCtrl::Create(wxWindow *parent,
     // UDM_SETBUDDY changes its size using some unknown algorithm, so setting
     // the sizes earlier is useless.
-    const int bestSpinWidth = wxSpinButton::DoGetBestSize().x;
-    const int effectiveSpinWidth = bestSpinWidth - GetOverlap();
```
I'm suspicious of this line disappearing, it looks like we don't take the overlap into account at all any more. Why?

thesiv replied:
Before my changes `effectiveSpinWidth` was used to calculate the width from the fixed value (`DEFAULT_ITEM_WIDTH`).
```
sizeCtrl.x = FromDIP(DEFAULT_ITEM_WIDTH) + effectiveSpinWidth;
```
But now we use our realization of `DoGetBestSize` (not `wxSpinButton::DoGetBestSize` as before), which uses `DoGetSizeFromTextSize`.

vadz commented on src/msw/spinctrl.cpp:
```diff
@@ -746,9 +769,18 @@ int wxSpinCtrl::GetOverlap() const
+    const int base = GetBase();
+    const int digitsCount = wxMax(GetDigitsCount(GetMin(), base),
+                                  GetDigitsCount(GetMax(), base));
+    return GetBestSizeFromDigitsCount(digitsCount);
```
Why do we have to go through "digits count"? Couldn't we just use `GetSizeFromTextSize(wxMax(GetTextExtent(wxString::Format("%d", GetMin())).x, GetTextExtent(wxString::Format("%d", GetMax())).x))`?

Or do what `wxSpinCtrlGTKBase::DoGetBestSize()` does (of course, this should be extracted into a common function in `wxSpinCtrlBase` then).

thesiv replied:
`wxSpinCtrl` under `MSW` is derived from `wxSpinButton` so the common parent for `wxSpinCtrlGTKBase` (`GTK`) and `wxSpinCtrl` (`MSW`) is `wxControl` but not `wxSpinCtrlBase`. So I get rid of math calculations and use string formatting locally in `wxSpinCtrl`.

vadz replied:
Yes, the proper solution for this would be to use CRTP, i.e. create `wxSpinCtrlIface<T>` deriving from `T` and redefine `wxSpinCtrlBase` as `wxSpinCtrlIface<wxControl>`, while inheriting `wxSpinCtrl` in wxMSW from `wxSpinCtrlIface<wxSpinButton>`, but it's probably not worth doing this right now.

If there is enough of code to reuse, you could easily do it in some inline function in `include/wx/private/spinctrl.h` however.

vadz reviewed (commented):
I think the digits stuff is unnecessary and we should just use the strings directly.

Also, please recheck/explain the changes to the overlap handling in the best size logic.

TIA!

thesiv committed:
fixup! Improve the best size determination for wxSpinCtrl

thesiv committed:
fixup! Prettify wxGridCellNumberEditor placement in the grid

thesiv committed:
Fix wxSpinCtrl under GTK

Fix the initial min size and the best size.

thesiv committed:
Add GetSizeFromText helper function to wxControlBase

Add the helper function that uses combination of GetSizeFromTextSize and
GetTextExtent which used together pretty often.

vadz commented on samples/widgets/spinbtn.cpp:
```diff
@@ -459,14 +459,8 @@ void SpinBtnWidgetsPage::OnButtonSetMinAndMax(wxCommandEvent& WXUNUSED(event))
-
-    m_spinctrl->SetMinSize(size);
-    m_spinctrl->SetSize(size);
-
```
Sorry, I don't really understand the meaning of this change, could you please explain?

vadz commented on src/gtk/spinctrl.cpp:
```diff
@@ -145,6 +145,10 @@ bool wxSpinCtrlGTKBase::Create(wxWindow *parent, wxWindowID id,
+    // Fix the min width because PostCreation sets it to `size`, but the
+    // control can be thinner.
+    SetMinSize(GetSizeFromText("9"));
+
```
No, this is wrong. The initial size must become the min size. If you don't want it to, don't create the control with this size.

But initial size is the min size by convention, if specified.

vadz commented on src/gtk/spinctrl.cpp:
```diff
@@ -354,15 +360,17 @@ GdkWindow *wxSpinCtrlGTKBase::GTKGetWindow(wxArrayGdkWindows& windows) const
 
     wxString longestText(wxMax(lenMin, lenMax), '9');
-    if ( minVal < 0 )
-        longestText.insert(0, "-");
```
It seems wrong to remove this.

thesiv replied:
Minus sign is handled by `Format`.

vadz commented on src/msw/spinctrl.cpp:
```diff
@@ -352,33 +352,32 @@ bool wxSpinCtrl::Create(wxWindow *parent,
+    // Fix the min width because SetInitialSize sets it to sizeCtrl, but the
+    // control can be thinner.
+    SetMinSize(GetSizeFromText("9"));
+
```
This is wrong too, for the same reason as in wxGTK.

vadz commented on src/msw/spinctrl.cpp:
```diff
@@ -439,10 +438,16 @@ bool wxSpinCtrl::SetBase(int base)
     // use ES_NUMBER for the buddy control.
     UpdateBuddyStyle();
 
+    // Update the control text.
```
I think this comment is useless, if it said "Update the displayed text after changing the base it uses", it would be better.

But I'm mostly interested in why do we do it in this order here, but in reverse order (`SetValue()` first, then `InvalidateBestSize()`) in wxGTK -- is there a good reason for this? If yes, could you please explain it in a comment? If not, could you please use the same order in both ports?

vadz reviewed (commented):
Sorry, there are still some questions/problems, at least with the initial size, which is not honoured any more: if I create `wxSpinCtrl` with 200px width, it really must be 200px.

thesiv committed:
fixup! Add GetSizeFromText helper function to wxControlBase

thesiv committed:
fixup! Improve the best size determination for wxSpinCtrl

thesiv committed:
fixup! Fix wxSpinCtrl under GTK

thesiv commented on src/msw/spinctrl.cpp:
```diff
@@ -748,7 +755,15 @@ int wxSpinCtrl::GetOverlap() const
+    const char* formatString = GetBase() == 16 ?
+                               (maxVal < 0x10000 ? "0x%04lx" : "0x%08lx") :
+                               "%d";
+    const int lenMin = wxString::Format(formatString, minVal).length();
```
Why?
```c++
wxString::Format("%d", -10)
```
Will return "-10" (length of 3).

thesiv commented:
Fixed.

vadz commented on include/wx/spinctrl.h:
```diff
@@ -143,6 +143,11 @@ namespace wxPrivate
 extern wxString wxSpinCtrlFormatAsHex(long val, long maxVal);
 
+// The helper function to determine the best size for the given control.
+// We can't implement this functional in the wxSpinCtrlBase because MSW implementation
```
```suggestion
// We can't implement this function in the wxSpinCtrlBase because MSW implementation
```

vadz commented on interface/wx/control.h:
```diff
@@ -140,6 +140,27 @@ class wxControl : public wxWindow
     wxSize GetSizeFromTextSize(const wxSize& tsize) const;
 
+    /**
+        Determine the size needed by the control to leave the given text.
```
```suggestion
        Determine the minimum size needed by the control to display the given text.
```

vadz commented on src/msw/spinctrl.cpp:
```diff
@@ -348,57 +348,56 @@ bool wxSpinCtrl::Create(wxWindow *parent,
     SetInitialSize(sizeCtrl);
 
-    (void)::ShowWindow(GetBuddyHwnd(), SW_SHOW);
+    // Fix the min size.
```
Can't we avoid calling `SetInitialSize()` here? It looks wrong to call it and then undo part of what it does...

If we really can't, the comment should explain better what's going on:
```suggestion
    // SetInitialSize() also sets the min size, but it should only be set to
    // the value specified by user, not the one computed by us, so override
    // the min size set just above.
```

vadz reviewed (commented):
I still have the question about `SetInitialSize()`, but if there is some problem with it, let's merge it like this.

thesiv committed:
fixup! Add GetSizeFromText helper function to wxControlBase

thesiv committed:
fixup! Improve the best size determination for wxSpinCtrl

thesiv commented:
Fixed.

vadz commented on src/msw/spinctrl.cpp:
```diff
@@ -385,17 +385,7 @@ bool wxSpinCtrl::Create(wxWindow *parent,
                    name, size.x, bestSize.GetWidth());
     }
```
`sizeCtrl` is unused now, right? So all this code could basically be replaced with just
```cpp
if ( size.x > 0 && size.x < GetBestSize().x )
{
     wxLogDebug(...);
}
```
couldn't it?

thesiv committed:
fixup! Improve the best size determination for wxSpinCtrl

thesiv commented:
Fixed.

thesiv committed:
fixup! Improve the best size determination for wxSpinCtrl

thesiv committed:
fixup! Improve the best size determination for wxSpinCtrl

vadz commented:
Thanks, rebased and merged now.
